### PR TITLE
Minor Plugin API Extensions

### PR DESCRIPTION
**Pull request by eplawless**
#### Overview

Hello! I'd like to make a few minor changes to the plugin API which will make it easier to keep the vanilla node-inspector source tree untouched (and up to date) and ship plugins alongside.

I'm using node-inspector plus some custom plugins to provide JS debugging support for the Netflix application on TVs and game consoles.

This change would make it so I can stop shipping a forked version :)

#### Proposed Changes
* Added --plugin-path argument for specifying root plugin path when plugins are enabled.
* Added plugin/manifest.json "override" property to allow finer-grained control of inspector.json.

### CODE REVIEW

**Comment by eplawless** [comment]
@3y3 Wondering if I could get your feedback on this change, whether it's missing anything etc.

**Comment by eplawless** [comment]
The reason for the "override" property in addition to "exclude" is in order to re-enable or tweak features that node-inspector has already set.

My immediate use case (likely with more to follow) is this plugin:

**manifest.json**
```json
{
  "name": "restore-elements-panel",
  "override": [{ "name": "elements", "type": "autostart" }]
}
```

**module.json**
```json
{}
```

**Comment by 3y3** [comment]
In current time I'm ready to accept this changes, but keep in mind, that plugins api is unofficial because it is uncompleted:
- [ ] DevTools Client
- [ ] Inspector process
- [ ] Application process
- [ ] Plugins installer

I will try to save ability to setup plugins dir, but I'm not sure, that there will not problems with `Plugins installer` step.
So I'll notify you about this as a first user of plugins api =)
And I'll call you in pull request for review.

I have some inline questions ip pr

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,5 @@ v8.log
 /test/fixtures/work
 npm-debug.log
 node-inspector-*.tgz
+*.swp
+*.swo
```
Comment (3y3) [inline .gitignore]: Unrelated change

File: lib/plugins.js [hunk 2]
```diff
@@ -2,8 +2,11 @@ var fs = require('fs'),
     inherits = require('util').inherits,
     path = require('path');
 
-var CWD = path.join(__dirname, '../plugins'),
-    THROW_CONFLICTS = true;
+var THROW_CONFLICTS = true;
+
+function getPluginPath(config) {
```
Comment (3y3) [inline lib/plugins.js]: Is it possible to save `CWD` option and rewrite in init function:

```js
var plugins = require('./plugins.js'); // ready to work with default parameters
plugins.init(options); // override CWD
```
?

File: lib/plugins.js [hunk 3]
```diff
@@ -30,18 +33,35 @@ function mergeByName(acceptor, donor, name, onConflict) {
...
-  dirlist = [];
+function _setMockPlugins(pluginPath, plugins) {
+  cachedPlugins[pluginPath] = plugins;
 }
-  
-dirlist.reduce(function(plugins, subdir) {
-    var _path = path.resolve(CWD, subdir, 'manifest.json'),
-        manifest;
+
+function getPlugins(config, ignoreCache) {
```
Comment (3y3) [inline lib/plugins.js]: There we use `ignoreCache ` with `ignoreCache => true`?

File: lib/debug-server.js [hunk 4]
```diff
@@ -119,7 +119,7 @@ DebugServer.prototype.start = function(options) {
   app.get('/', debugAction.bind(this));
   app.get('/debug', redirectToRoot.bind(this));
   app.use('/node', express.static(OVERRIDES));
-  app.use('/plugins', express.static(PLUGINS));
+  app.use('/plugins', express.static(getPluginPath(options)));
```
Comment (3y3) [inline lib/debug-server.js]: Why we need to recalculate plugins on each request to them?

File: lib/plugins.js (hunk 3 above)
Comment (eplawless) [inline lib/plugins.js]: Yes, you could call `getPlugins({ plugins: true, pluginPath: '/foo/bar' }, true)` and it will ignore the cached version and read from the directory again.

I added a separate argument so people don't have to extend or mutate the config object.

**Comment by 3y3** [comment]
Please isolate `override` option in new commit in current pr.

File: lib/plugins.js (hunk 2 above)
Comment (eplawless) [inline lib/plugins.js]: Yeah, I could definitely make that change if you like.

File: .gitignore (hunk 1 above)
Comment (eplawless) [inline .gitignore]: I'll remove it from the PR.

**Comment by eplawless** [comment]
Sounds good, I'll make those changes.

File: lib/plugins.js (hunk 3 above)
Comment (3y3) [inline lib/plugins.js]: Yes, I understand how it can be used, but I don't see where we use this logic.

I'm not sure that reinitialisation of plugins on the fly is good think. It's useful in plugins development, but for end user this can cause some unexpected problems. Can we avoid caching this logic on firs step?

Comment (eplawless) [inline lib/plugins.js]: I can definitely remove it, it's unused currently.

**Comment by eplawless** [comment]
Made those edits, let me know if I missed anything?

File: lib/plugins.js (hunk 2 above)
Comment (eplawless) [inline lib/plugins.js]: It's used in lib/plugins.js now, but not in test/plugins.js; I can modify the test file as well, if you like.

**Comment by eplawless** [comment]
We're going to be actively developing and maintaining plugins, so I'm perfectly happy to adapt to the plugins API as it evolves and changes.  I think it's going to enable some really neat tools for our teams.

If you need some help with node-inspector itself at any point, I think I can swing some dedicated time for contributing to it.

**Comment by eplawless** [comment]
@3y3 is there anything else I should change?

**Comment by 3y3** [comment]
No. All ready and merged. I need to fix some bugs before next release, so wait next release after weekend.

Thank you.
